feat(internal): Add `kraft cloud` services, volumes, and quotas support

craciunoiuc opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [x] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

~This will need rebasing once #1010 is in to remove the first 2 commits.~

GitHub-Depends: #1010

craciunoiuc committed:
feat!: Bump `kraft cloud` commands to latest SDK (v0.4.0)

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(kraft): Introduce `kraft cloud volume` subcommands

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(kraft): Introduce `kraft cloud service` subcommands

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(kraft): Introduce `kraft cloud quotas` subcommand

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

nderjung commented on internal/cli/kraft/cloud/quotas/quotas.go:
```diff
@@ -0,0 +1,78 @@
+		Short:   "View your resource quota on KraftCloud",
+		Use:     "quotas",
+		Aliases: []string{"q", "quota"},
+		Annotations: map[string]string{
+			cmdfactory.AnnotationHelpGroup: "kraftcloud-quota",
+		},
```
Since there are no subcommands, move this to the parent group
```diff
  		Annotations: map[string]string{
- 			cmdfactory.AnnotationHelpGroup: "kraftcloud-quota",
+ 			cmdfactory.AnnotationHelpGroup: "kraftcloud",
  		},
```

nderjung commented on internal/cli/kraft/cloud/cloud.go:
```diff
@@ -81,6 +84,15 @@ func NewCmd() *cobra.Command {
+	cmd.AddGroup(&cobra.Group{ID: "kraftcloud-svc", Title: "SERVICE COMMANDS"})
+	cmd.AddCommand(service.NewCmd())
+
+	cmd.AddGroup(&cobra.Group{ID: "kraftcloud-quota", Title: "QUOTA COMMANDS"})
```
Since quotas does not have subsequent subcommands, this group can be removed:
```diff
- cmd.AddGroup(&cobra.Group{ID: "kraftcloud-quota", Title: "QUOTA COMMANDS"})
```
Move the `AddCommand` to be under the one for deploy.

nderjung reviewed (commented):
For the ability to show the new `kraft cloud` subcommands in the main help menu, simply add the groups to the top-level command, i.e. edit `internal/cli/kraft/kraft.go` and add:
```
	cmd.AddGroup(&cobra.Group{ID: "kraftcloud-vol", Title: "KRAFT CLOUD VOLUME COMMANDS"})
	cmd.AddGroup(&cobra.Group{ID: "kraftcloud-svc", Title: "KRAFT CLOUD SERVICE GROUP COMMANDS"})
```

nderjung commented on internal/cli/kraft/cloud/instance/get/get.go:
```diff
@@ -21,23 +21,23 @@ import (
-	cmd, err := cmdfactory.New(&StatusOptions{}, cobra.Command{
+	cmd, err := cmdfactory.New(&GetOptions{}, cobra.Command{
 		Short:   "Retrieve the status of an instance",
 		Use:     "status [FLAGS] UUID",
```
Don't forget to switch the `Use` and add `status` as an alias:
```diff
diff --git a/internal/cli/kraft/cloud/instance/get/get.go b/internal/cli/kraft/cloud/instance/get/get.go
index 1573801..fce143a 100644
--- a/internal/cli/kraft/cloud/instance/get/get.go
+++ b/internal/cli/kraft/cloud/instance/get/get.go
@@ -39,9 +39,9 @@ func Get(ctx context.Context, opts *GetOptions, args ...string) error {
 func NewCmd() *cobra.Command {
        cmd, err := cmdfactory.New(&GetOptions{}, cobra.Command{
                Short:   "Retrieve the status of an instance",
-               Use:     "status [FLAGS] UUID",
+               Use:     "get [FLAGS] UUID",
                Args:    cobra.ExactArgs(1),
-               Aliases: []string{"info"},
+               Aliases: []string{"status", "info"},
                Example: heredoc.Doc(`
                        # Retrieve information about a kraftcloud instance
                        $ kraft cloud instance status fd1684ea-7970-4994-92d6-61dcc7905f2b
```

nderjung commented on internal/cli/kraft/cloud/utils/print.go:
```diff
@@ -46,7 +49,7 @@ func PrintInstances(ctx context.Context, format string, instances ...kraftcloudi
 	if format != "table" {
```
Let's add the newly exposed `NAME` field:
```diff
diff --git a/internal/cli/kraft/cloud/utils/print.go b/internal/cli/kraft/cloud/utils/print.go
index e368e06..b792070 100644
--- a/internal/cli/kraft/cloud/utils/print.go
+++ b/internal/cli/kraft/cloud/utils/print.go
@@ -45,6 +45,7 @@ func PrintInstances(ctx context.Context, format string, instances ...kraftcloudi
        if format != "table" {
                table.AddField("UUID", cs.Bold)
        }
+       table.AddField("NAME", cs.Bold)
        table.AddField("DNS", cs.Bold)
        if format != "table" {
                table.AddField("PRIVATE IP", cs.Bold)
@@ -72,6 +73,7 @@ func PrintInstances(ctx context.Context, format string, instances ...kraftcloudi
                if format != "table" {
                        table.AddField(instance.UUID, nil)
                }
+               table.AddField(instance.Name, nil)
                table.AddField(instance.FQDN, nil)
                if format != "table" {
                        table.AddField(instance.PrivateIP, nil)
```

nderjung commented on internal/cli/kraft/cloud/deploy/deploy.go:
```diff
@@ -40,7 +40,7 @@ type DeployOptions struct {
+	Replicas  int                       `local:"true" long:"replicas" short:"R" usage:"Number of replicas of the instance" default:"0"`
 	Strategy  packmanager.MergeStrategy `noattribute:"true"`
 	Timeout   time.Duration             `local:"true" long:"timeout" usage:"Set the timeout for remote procedure calls"`
 	Workdir   string                    `local:"true" long:"workdir" short:"w" usage:"Set an alternative working directory (default is cwd)"`
```
The API/SDK now properly outputs digest and tag syntaxes as-per the OCI distribution spec, this means:
```diff
diff --git a/internal/cli/kraft/cloud/deploy/deploy.go b/internal/cli/kraft/cloud/deploy/deploy.go
index aa02861..c0f04d3 100644
--- a/internal/cli/kraft/cloud/deploy/deploy.go
+++ b/internal/cli/kraft/cloud/deploy/deploy.go
@@ -209,7 +209,7 @@ func (opts *DeployOptions) Run(ctx context.Context, args []string) error {
                                        }
 
                                        for _, image := range images {
-                                               split := strings.Split(image.Digest, "/")
+                                               split := strings.Split(image.Digest, "@sha256:")
                                                if !strings.HasPrefix(split[len(split)-1], digest) {
                                                        continue
                                                }
```

nderjung commented on internal/cli/kraft/cloud/service/list/list.go:
```diff
@@ -0,0 +1,82 @@
+
+	serviceGroups, err := client.WithMetro(opts.metro).List(ctx)
+	if err != nil {
+		return fmt.Errorf("could not list volumes: %w", err)
```
```diff
- return fmt.Errorf("could not list volumes: %w", err)
+ return fmt.Errorf("could not list service groups: %w", err)
```

nderjung commented on internal/cli/kraft/cloud/utils/print.go:
```diff
@@ -69,21 +73,204 @@ func PrintInstances(ctx context.Context, format string, instances ...kraftcloudi
+			for _, handler := range service.Handlers {
+				handlers = append(handlers, string(handler))
+			}
+			services = append(services, fmt.Sprintf("%d:%d:%s", service.Port, service.DestinationPort, strings.Join(handlers, ",")))
```
To be consistent with how we parse `-p` from the CLI:
```diff
diff --git a/internal/cli/kraft/cloud/utils/print.go b/internal/cli/kraft/cloud/utils/print.go
index b792070..ad19d46 100644
--- a/internal/cli/kraft/cloud/utils/print.go
+++ b/internal/cli/kraft/cloud/utils/print.go
@@ -206,7 +206,7 @@ func PrintServiceGroups(ctx context.Context, format string, serviceGroups ...kra
                        for _, handler := range service.Handlers {
                                handlers = append(handlers, string(handler))
                        }
-                       services = append(services, fmt.Sprintf("%d:%d:%s", service.Port, service.DestinationPort, strings.Join(handlers, ",")))
+                       services = append(services, fmt.Sprintf("%d:%d/%s", service.Port, service.DestinationPort, strings.Join(handlers, "+")))
                }
                table.AddField(strings.Join(services, ", "), nil)
                table.AddField(strings.Join(sg.Instances, ","), nil)
```

nderjung committed:
feat(utils): Introduce method `IsUUID`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(cloud): Introduce `PrettyPrintInstance`

This method is abstracted away from `kraft cloud deploy` and
introduced also in `kraft cloud instance create`.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(img): Introduce `kraft cloud img remove`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(runtime): Introduce `SetName` to overwrite the runtime name

Signed-off-by: Alexander Jung <alex@unikraft.io>
